### PR TITLE
feature: support relativePosition.radius in subscriptions and REST requests

### PR DESCRIPTION
**Pull request by sbender9**
This feature allows ws subscriptions to be based on the relative position of the vessel. This allows the client to to only get information about other vessels that are within a specific distance. For example:

```
{
  context: {
    radius: 10,
    position: {
      longitude: -76.4639314,
      latitude: 39.0700403
    }
  },
  subscribe: [
    {
      path: 'navigation.*'
    }
  ]
}
```

### CODE REVIEW

File: lib/interfaces/rest.js [hunk 1]
```diff
@@ -71,15 +73,45 @@ module.exports = function (app) {
...
+
+            if (position) {
+              if (
+                geolib.isPointInCircle(
+                  position.value,
+                  req.query,
+                  req.query.radius
+                )
+              ) {
+                debug(`vessel: ${key} ${JSON.stringify(position)}`)
```
Comment (tkurki) [inline lib/interfaces/rest.js:98]: Very minor here, but as discussed in canboatjs JSON.stringify is surprisingly heavy and I'd like to start the habit of doing this like `debug.enabled && debug(...)`

File: lib/interfaces/ws.js [hunk 2]
```diff
@@ -228,3 +238,16 @@ function normalizeUpdate (update) {
     }
   })
 }
+
+function updatePosition (delta, spark) {
```
Comment (tkurki) [inline lib/interfaces/ws.js]: change this to `updatePosition(delta, position)` as this doesn't need to know anything about spark.request?

File: lib/interfaces/ws.js [hunk 3]
```diff
@@ -97,9 +97,19 @@ module.exports = function (app) {
...
+            // if this has a relativePosition subscription, update the position
+            updatePosition(msg, spark)
+          }
           app.handleMessage(spark.request.source || 'ws', msg)
         }
         if (msg.subscribe) {
+          if (!_.isUndefined(msg.context.relativePosition)) {
+            debug('setting up to store position')
+            spark.request.position = {}
+            msg.context.relativePosition.position = spark.request.position
```
Comment (tkurki) [inline lib/interfaces/ws.js]: `relativePosition` is not very clear, how about `msg.context.subscriber.position`? Ditto for `spark.request.subscriberPosition`.

File: lib/subscriptionmanager.js [hunk 4]
```diff
@@ -153,16 +159,29 @@ function pathMatcher (path) {
...
+          subscribeCommand.context === 'self') &&
+          normalizedDeltaData.context === selfContext)
+    } else if (_.get(subscribeCommand.context, 'relativePosition')) {
+      var relativePosition = subscribeCommand.context.relativePosition
+      if (!_.get(relativePosition, 'radius')) {
+        errorCallback('Please specify a radius for relativePosition')
+        return x => false
+      }
+      return normalizedDeltaData => {
+        return checkPosition(app, relativePosition, normalizedDeltaData)
```
Comment (tkurki) [inline lib/subscriptionmanager.js]: Doesn't checkPosition here get the relativePosition from the subscribeCommand, which is static and will never change? So we are updating the subscriber's position in spark.request, but that is never used in checkPosition?

Thinking this through how useful is the updating current position going to be in real life? It only has the lifetime of the ws connection and can be easily updated by hanging off the ws conenction and recconnecting every x hours. For this to be useful we would need to be dealing with orbital speeds or very small radii. What do you think? Ditch or keep?

Comment (sbender9) [reply lib/subscriptionmanager.js]: Yep. You’re right. I’ll ditch it

**Comment by sbender9** [comment]
I think the only thing left here is the naming. One option for the subscription:

```
{
          context: {
              contexts: "vessels.*",
              radius: 10,
              position: {
                longitude: -76.4639314,
                latitude: 39.0700403
              }
            }
          },
          subscribe: [
            {
              path: '*'
            }
          ]
        }
```

**Comment by sbender9** [comment]
Or just

```
{
  context: {
    radius: 1,
    position: {
      longitude: -76.4639314,
      latitude: 39.0700403
    }
  },
  subscribe: [
    {
      path: 'navigation.position'
    }
  ]
}

```

**Pull request merged**